### PR TITLE
fix: simplify `ListOffsetArray_reduce_nonlocal_outstartsstops`

### LINKED ISSUE #1784: Segmentation fault with nested pair-wise comparison
### Version of Awkward Array

1.9.0 and 1.10.0

### Description and code to reproduce

Used as part of coffea package: 

```python
from coffea.nanoevents import NanoEventsFactory
from coffea.nanoevents.schemas import TreeMakerSchema
import awkward as ak

events = NanoEventsFactory.from_root(
    "root://cmseos.fnal.gov//store/user/lpcsusyhad/ExoEMJAnalysis2020/Run2ProductionV20Skimmed/Summer20UL16/GJets_DR-0p4_HT-100To200_TuneCP5_13TeV-madgraphMLM-pythia8/0_RA2AnalysisTree.root",
    schemaclass=TreeMakerSchema,
    treepath="TreeMaker2/PreSelection",
    entry_start=10000,
    entry_stop=20000).events()

jets = events.Jets[events.Jets.pt > 100]
tracks = events.Tracks[events.Tracks.pt > 1.0]

_, tracks = ak.unzip(ak.cartesian([jets.eta, tracks], nested=True))
tracks['deltaR'] = jets.delta_r(tracks)

t_index = ak.argmin(tracks.deltaR, axis=-2)
j_index = ak.local_index(jets.pt, axis=-1)
_, t_index = ak.unzip(ak.cartesian([j_index, t_index], nested=True))
tracks = tracks[j_index == t_index]

tracks = tracks[tracks.deltaR < 0.4]
```

The above segmentation faults at different point for different input files. Usually around the `ak.cartesian` calls, but also when adding variables to the array. Segmentation fault was observed for both awkward 1.9.0 and awkward 1.10.0, but never occurred for awkward 1.8.0

### PR DESCRIPTION
**Pull request by agoose77**
- [x] fixes #1784
- [x] fixeds #1791
- [x] removes unused kernel
- [x] adds test

This PR fixes #1784 by rewriting the `awkward_ListOffsetArray_reduce_nonlocal_outstartsstops_64` kernel. The actual fix for #1784 is just to wrap the entire kernel body in the `if (distincts[i] != -1)` guard. However, I took the opportunity to simplify and explain the kernel. This means that we can drop the `gaps` kernel, and we don't need to perform any float divisions. So, I'd hope that this is also a slight performance boost (but this did not motivate the rewrite).

I *hope* that I haven't overlooked any design invariants or corner-cases :crossed_fingers:; I've attempted to understand the purpose and function of the kernel in depth before writing this PR.

:warning: I did change the expected outcomes of the test suite - some of the kernel tests check the offset values. We don't care about the actual values of `starts` and `stops` as long as they're in-bounds and equal.

### CODE REVIEW

File: src/cpu-kernels/awkward_ListOffsetArray_reduce_nonlocal_outstartsstops_64.cpp [hunk 1]
```diff
@@ -1,51 +1,47 @@
 // BSD 3-Clause License; see https://github.com/scikit-hep/awkward-1.0/blob/main/LICENSE
 
-#define FILENAME(line) FILENAME_FOR_EXCEPTIONS_C("src/cpu-kernels/awkward_ListOffsetArray_reduce_nonlocal_outstartsstops_64.cpp", line)
-
+#define FILENAME(line)                                                 \
+  FILENAME_FOR_EXCEPTIONS_C(                                           \
+      "src/cpu-kernels/"                                               \
+      "awkward_ListOffsetArray_reduce_nonlocal_outstartsstops_64.cpp", \
+      line)
```
Comment (jpivarski) [inline src/cpu-kernels/awkward_ListOffsetArray_reduce_nonlocal_outstartsstops_64.cpp:7]: These were all in one line because they're easier to maintain with text-based tools (`awk`, `sed`, `grep`), but we've never needed to do that after their initial generation, so I doubt there will be any issues with this reformatting.

**Review by jpivarski — APPROVED** [review]
I'm surprised that it's possible to solve this problem without the `gaps`, though `gaps` was added to handle a special case, after the original algorithm was first written. There are probably tests for it in the suite, but it's hard to tell which ones would be relevant, so I wrote a new test that is sensitive to what the `gaps` are trying to accomplish.

First, I _found_ a test that has non-trivial `gaps`. Then I turned it into a commandline example for flexibility. Everything between the next two horizontal lines is in `main` (with `gaps`), to get an unchanged baseline. The commandline prints the original array layout followed by the computed output.

I modified the code to print the `gaps` on line `listoffsetarray.py:1510`, right after they are defined. The size of the `gaps` array is large enough for the _maximum_ that `gaps` can be, so (in this example) the last item is undefined.

------------

Product of irregular prime numbers, two ListOffsetArrays deep.

```
% python -c 'import awkward as ak, numpy as np; array = ak.Array(ak.contents.ListOffsetArray(ak.index.Index64([0, 4, 4, 6]), ak.contents.ListOffsetArray(ak.index.Index64([0, 3, 3, 5, 6, 8, 9]), ak.contents.NumpyArray([ 2,  3,  5,  7, 11, 13, 17, 19, 23])))); print(array.layout); print(ak.Array(ak.prod(array, axis=-2), check_valid=True))'

<ListOffsetArray len='3'>
    <offsets><Index dtype='int64' len='4'>[0 4 4 6]</Index></offsets>
    <content><ListOffsetArray len='6'>
        <offsets><Index dtype='int64' len='7'>[0 3 3 5 6 8 9]</Index></offsets>
        <content><NumpyArray dtype='int64' len='9'>[ 2  3  5  7 11 13 17 19 23]</NumpyArray></content>
    </ListOffsetArray></content>
</ListOffsetArray>
gaps [                  1                   2 4607182418800017408]
[[182, 33, 5], [], [391, 19]]
```

(`[[182, 33, 5], [], [391, 19]]` is the correct answer.)

Now the ListOffsetArrays don't start at 0. There's extra junk in the contents before the real lists start.

```
% python -c 'import awkward as ak, numpy as np; array = ak.Array(ak.contents.ListOffsetArray(ak.index.Index64(np.array([0, 4, 4, 6]) + 2), ak.contents.ListOffsetArray(ak.index.Index64(np.array([123, 321, 0, 3, 3, 5, 6, 8, 9]) + 1), ak.contents.NumpyArray([9999, 2,  3,  5,  7, 11, 13, 17, 19, 23])))); print(array.layout); print(ak.Array(ak.prod(array, axis=-2), check_valid=True))'

<ListOffsetArray len='3'>
    <offsets><Index dtype='int64' len='4'>[2 6 6 8]</Index></offsets>
    <content><ListOffsetArray len='8'>
        <offsets><Index dtype='int64' len='9'>
            [124 322   1   4   4   6   7   9  10]
        </Index></offsets>
        <content><NumpyArray dtype='int64' len='10'>
            [9999    2    3    5    7   11   13   17   19   23]
        </NumpyArray></content>
    </ListOffsetArray></content>
</ListOffsetArray>
gaps [                  1                   2 4607182418800017408]
[[182, 33, 5], [], [391, 19]]
```

Now they're ListArrays, rather than ListOffsetArrays.

```
% python -c 'import awkward as ak, numpy as np; array = ak.Array(ak.contents.ListArray(ak.index.Index64([0, 4, 4]), ak.index.Index64([4, 4, 6]), ak.contents.ListArray(ak.index.Index64([0, 3, 3, 5, 6, 8]), ak.index.Index64([3, 3, 5, 6, 8, 9]), ak.contents.NumpyArray([ 2,  3,  5,  7, 11, 13, 17, 19, 23])))); print(array.layout); print(ak.Array(ak.prod(array, axis=-2), check_valid=True))'

<ListArray len='3'>
    <starts><Index dtype='int64' len='3'>[0 4 4]</Index></starts>
    <stops><Index dtype='int64' len='3'>[4 4 6]</Index></stops>
    <content><ListArray len='6'>
        <starts><Index dtype='int64' len='6'>[0 3 3 5 6 8]</Index></starts>
        <stops><Index dtype='int64' len='6'>[3 3 5 6 8 9]</Index></stops>
        <content><NumpyArray dtype='int64' len='9'>[ 2  3  5  7 11 13 17 19 23]</NumpyArray></content>
    </ListArray></content>
</ListArray>
gaps [                  1                   2 4612217596255138984]
[[182, 33, 5], [], [391, 19]]
```

Now the cases in which `starts[i] == stops[i]` are replaced with crazy values, which is legal. (If `starts[i] == stops[i]`, the list is considered empty. Neither `starts[i]` nor `stops[i]` have to be sane.)

```
% python -c 'import awkward as ak, numpy as np; array = ak.Array(ak.contents.ListArray(ak.index.Index64([0, 100, 4]), ak.index.Index64([4, 100, 6]), ak.contents.ListArray(ak.index.Index64([0, 1000, 3, 5, 6, 8]), ak.index.Index64([3, 1000, 5, 6, 8, 9]), ak.contents.NumpyArray([ 2,  3,  5,  7, 11, 13, 17, 19, 23])))); print(array.layout); print(ak.Array(ak.prod(array, axis=-2), check_valid=True))'

<ListArray len='3'>
    <starts><Index dtype='int64' len='3'>[  0 100   4]</Index></starts>
    <stops><Index dtype='int64' len='3'>[  4 100   6]</Index></stops>
    <content><ListArray len='6'>
        <starts><Index dtype='int64' len='6'>
            [   0 1000    3    5    6    8]
        </Index></starts>
        <stops><Index dtype='int64' len='6'>
            [   3 1000    5    6    8    9]
        </Index></stops>
        <content><NumpyArray dtype='int64' len='9'>[ 2  3  5  7 11 13 17 19 23]</NumpyArray></content>
    </ListArray></content>
</ListArray>
gaps [                  1                   2 4612217596255138984]
[[182, 33, 5], [], [391, 19]]
```

Now there's junk data between some of the valid items, but the `starts` and `stops` skip it.

```
% python -c 'import awkward as ak, numpy as np; array = ak.Array(ak.contents.ListArray(ak.index.Index64([0, 100, 6]), ak.index.Index64([4, 100, 8]), ak.contents.ListArray(ak.index.Index64([0, 1000, 3, 5, 60, 80, 6, 8]), ak.index.Index64([3, 1000, 5, 6, 80, 90, 8, 9]), ak.contents.NumpyArray([ 2,  3,  5,  7, 11, 13, 17, 19, 23])))); print(array.layout); print(ak.Array(ak.prod(array, axis=-2), check_valid=True))'

<ListArray len='3'>
    <starts><Index dtype='int64' len='3'>[  0 100   6]</Index></starts>
    <stops><Index dtype='int64' len='3'>[  4 100   8]</Index></stops>
    <content><ListArray len='8'>
        <starts><Index dtype='int64' len='8'>
            [   0 1000    3    5   60   80    6    8]
        </Index></starts>
        <stops><Index dtype='int64' len='8'>
            [   3 1000    5    6   80   90    8    9]
        </Index></stops>
        <content><NumpyArray dtype='int64' len='9'>[ 2  3  5  7 11 13 17 19 23]</NumpyArray></content>
    </ListArray></content>
</ListArray>
gaps [                  1                   2 4612217596255138984]
[[182, 33, 5], [], [391, 19]]
```

Now one of the items is actually out of order (the first 3 items of the NumpyArray have been moved to the end).

```
% python -c 'import awkward as ak, numpy as np; array = ak.Array(ak.contents.ListArray(ak.index.Index64([0, 100, 6]), ak.index.Index64([4, 100, 8]), ak.contents.ListArray(ak.index.Index64([9, 1000, 3, 5, 60, 80, 6, 8]), ak.index.Index64([12, 1000, 5, 6, 80, 90, 8, 9]), ak.contents.NumpyArray([ 20,  30,  50,  7, 11, 13, 17, 19, 23, 2, 3, 5])))); print(array.layout); print(ak.Array(ak.prod(array, axis=-2), check_valid=True))'

<ListArray len='3'>
    <starts><Index dtype='int64' len='3'>[  0 100   6]</Index></starts>
    <stops><Index dtype='int64' len='3'>[  4 100   8]</Index></stops>
    <content><ListArray len='8'>
        <starts><Index dtype='int64' len='8'>
            [   9 1000    3    5   60   80    6    8]
        </Index></starts>
        <stops><Index dtype='int64' len='8'>
            [  12 1000    5    6   80   90    8    9]
        </Index></stops>
        <content><NumpyArray dtype='int64' len='12'>
            [20 30 50  7 11 13 17 19 23  2  3  5]
        </NumpyArray></content>
    </ListArray></content>
</ListArray>
gaps [                  1                   2 4612217596255138984]
[[182, 33, 5], [], [391, 19]]
```

In each of the above cases, the high-level view of the input array is the same, though the layout differs.

---------

Okay, now same thing, but in this PR's branch. I removed the print-out in Python and added one to the kernel code to be sure that I'm running the new kernel code. (Everything was recompiled.)

```
% python -c 'import awkward as ak, numpy as np; array = ak.Array(ak.contents.ListOffsetArray(ak.index.Index64([0, 4, 4, 6]), ak.contents.ListOffsetArray(ak.index.Index64([0, 3, 3, 5, 6, 8, 9]), ak.contents.NumpyArray([ 2,  3,  5,  7, 11, 13, 17, 19, 23])))); print(array.layout); print(ak.Array(ak.prod(array, axis=-2), check_valid=True))'

<ListOffsetArray len='3'>
    <offsets><Index dtype='int64' len='4'>[0 4 4 6]</Index></offsets>
    <content><ListOffsetArray len='6'>
        <offsets><Index dtype='int64' len='7'>[0 3 3 5 6 8 9]</Index></offsets>
        <content><NumpyArray dtype='int64' len='9'>[ 2  3  5  7 11 13 17 19 23]</NumpyArray></content>
    </ListOffsetArray></content>
</ListOffsetArray>
no gaps
[[182, 33, 5], [], [391, 19]]
```

```
% python -c 'import awkward as ak, numpy as np; array = ak.Array(ak.contents.ListOffsetArray(ak.index.Index64(np.array([0, 4, 4, 6]) + 2), ak.contents.ListOffsetArray(ak.index.Index64(np.array([123, 321, 0, 3, 3, 5, 6, 8, 9]) + 1), ak.contents.NumpyArray([9999, 2,  3,  5,  7, 11, 13, 17, 19, 23])))); print(array.layout); print(ak.Array(ak.prod(array, axis=-2), check_valid=True))'

<ListOffsetArray len='3'>
    <offsets><Index dtype='int64' len='4'>[2 6 6 8]</Index></offsets>
    <content><ListOffsetArray len='8'>
        <offsets><Index dtype='int64' len='9'>
            [124 322   1   4   4   6   7   9  10]
        </Index></offsets>
        <content><NumpyArray dtype='int64' len='10'>
            [9999    2    3    5    7   11   13   17   19   23]
        </NumpyArray></content>
    </ListOffsetArray></content>
</ListOffsetArray>
no gaps
[[182, 33, 5], [], [391, 19]]
```

```
% python -c 'import awkward as ak, numpy as np; array = ak.Array(ak.contents.ListArray(ak.index.Index64([0, 4, 4]), ak.index.Index64([4, 4, 6]), ak.contents.ListArray(ak.index.Index64([0, 3, 3, 5, 6, 8]), ak.index.Index64([3, 3, 5, 6, 8, 9]), ak.contents.NumpyArray([ 2,  3,  5,  7, 11, 13, 17, 19, 23])))); print(array.layout); print(ak.Array(ak.prod(array, axis=-2), check_valid=True))'

<ListArray len='3'>
    <starts><Index dtype='int64' len='3'>[0 4 4]</Index></starts>
    <stops><Index dtype='int64' len='3'>[4 4 6]</Index></stops>
    <content><ListArray len='6'>
        <starts><Index dtype='int64' len='6'>[0 3 3 5 6 8]</Index></starts>
        <stops><Index dtype='int64' len='6'>[3 3 5 6 8 9]</Index></stops>
        <content><NumpyArray dtype='int64' len='9'>[ 2  3  5  7 11 13 17 19 23]</NumpyArray></content>
    </ListArray></content>
</ListArray>
no gaps
[[182, 33, 5], [], [391, 19]]
```

```
% python -c 'import awkward as ak, numpy as np; array = ak.Array(ak.contents.ListArray(ak.index.Index64([0, 100, 4]), ak.index.Index64([4, 100, 6]), ak.contents.ListArray(ak.index.Index64([0, 1000, 3, 5, 6, 8]), ak.index.Index64([3, 1000, 5, 6, 8, 9]), ak.contents.NumpyArray([ 2,  3,  5,  7, 11, 13, 17, 19, 23])))); print(array.layout); print(ak.Array(ak.prod(array, axis=-2), check_valid=True))'

<ListArray len='3'>
    <starts><Index dtype='int64' len='3'>[  0 100   4]</Index></starts>
    <stops><Index dtype='int64' len='3'>[  4 100   6]</Index></stops>
    <content><ListArray len='6'>
        <starts><Index dtype='int64' len='6'>
            [   0 1000    3    5    6    8]
        </Index></starts>
        <stops><Index dtype='int64' len='6'>
            [   3 1000    5    6    8    9]
        </Index></stops>
        <content><NumpyArray dtype='int64' len='9'>[ 2  3  5  7 11 13 17 19 23]</NumpyArray></content>
    </ListArray></content>
</ListArray>
no gaps
[[182, 33, 5], [], [391, 19]]
```

```
% python -c 'import awkward as ak, numpy as np; array = ak.Array(ak.contents.ListArray(ak.index.Index64([0, 100, 6]), ak.index.Index64([4, 100, 8]), ak.contents.ListArray(ak.index.Index64([0, 1000, 3, 5, 60, 80, 6, 8]), ak.index.Index64([3, 1000, 5, 6, 80, 90, 8, 9]), ak.contents.NumpyArray([ 2,  3,  5,  7, 11, 13, 17, 19, 23])))); print(array.layout); print(ak.Array(ak.prod(array, axis=-2), check_valid=True))'

<ListArray len='3'>
    <starts><Index dtype='int64' len='3'>[  0 100   6]</Index></starts>
    <stops><Index dtype='int64' len='3'>[  4 100   8]</Index></stops>
    <content><ListArray len='8'>
        <starts><Index dtype='int64' len='8'>
            [   0 1000    3    5   60   80    6    8]
        </Index></starts>
        <stops><Index dtype='int64' len='8'>
            [   3 1000    5    6   80   90    8    9]
        </Index></stops>
        <content><NumpyArray dtype='int64' len='9'>[ 2  3  5  7 11 13 17 19 23]</NumpyArray></content>
    </ListArray></content>
</ListArray>
no gaps
[[182, 33, 5], [], [391, 19]]
```

```
% python -c 'import awkward as ak, numpy as np; array = ak.Array(ak.contents.ListArray(ak.index.Index64([0, 100, 6]), ak.index.Index64([4, 100, 8]), ak.contents.ListArray(ak.index.Index64([9, 1000, 3, 5, 60, 80, 6, 8]), ak.index.Index64([12, 1000, 5, 6, 80, 90, 8, 9]), ak.contents.NumpyArray([ 20,  30,  50,  7, 11, 13, 17, 19, 23, 2, 3, 5])))); print(array.layout); print(ak.Array(ak.prod(array, axis=-2), check_valid=True))'

<ListArray len='3'>
    <starts><Index dtype='int64' len='3'>[  0 100   6]</Index></starts>
    <stops><Index dtype='int64' len='3'>[  4 100   8]</Index></stops>
    <content><ListArray len='8'>
        <starts><Index dtype='int64' len='8'>
            [   9 1000    3    5   60   80    6    8]
        </Index></starts>
        <stops><Index dtype='int64' len='8'>
            [  12 1000    5    6   80   90    8    9]
        </Index></stops>
        <content><NumpyArray dtype='int64' len='12'>
            [20 30 50  7 11 13 17 19 23  2  3  5]
        </NumpyArray></content>
    </ListArray></content>
</ListArray>
no gaps
[[182, 33, 5], [], [391, 19]]
```

Awesome!

------------------

In retrospect, these all had to work because all of the layouts are normalized through `toListOffsetArray64(True)` before the calculation with the `gaps` would have begun. So this was actually testing the `toListOffsetArray64(True)` function, but that's a prerequisite for the code you changed.

Anyway, even if these 6 tests were really the same test of `gaps`, we do know that it has been tested with non-trivial `gaps`. So I don't see how there could be any other corner cases to check.

I'll just add them to the test suite and enable auto-merge.

**Comment by jpivarski** [comment]
> ⚠️ I did change the expected outcomes of the test suite - some of the kernel tests check the offset values.

Those test outcomes were set by running the code itself, so the new values are just as good as the old ones. They were specifically the auto-generated unit tests, not the high-level tests we write by hand in the `tests` directory.
